nixos/lib/test-driver/lib/driver.py

```diff
@@ -0,0 +1,256 @@
+        vde_socket = tempfile.mkdtemp(
+            prefix="nixos-test-vde-", suffix="-vde{}.ctl".format(vlan_nr)
+        )
+        pty_master, pty_slave = pty.openpty()
+        vde_process = subprocess.Popen(
+            ["vde_switch", "-s", vde_socket, "--dirmode", "0700"],
+            stdin=pty_slave,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=False,
+        )
+        fd = os.fdopen(pty_master, "w")
+        fd.write("version\n")
+        # TODO: perl version checks if this can be read from
+        # an if not, dies. we could hang here forever. Fix it.
+        assert vde_process.stdout is not None
+        vde_process.stdout.readline()
+        if not os.path.exists(os.path.join(vde_socket, "ctl")):
+            raise Exception("cannot start vde_switch")
+
+        return (vlan_nr, vde_socket, vde_process, fd)
+
+
+class Logger:
+    def __init__(self) -> None:
+        self.enable_serial_logs = True  # switchable at runtime
+
+    def release(self) -> None:
+        pass
+
+    def log_machinestate(self, message: str, _: Dict[str, str] = {}) -> None:
+        print(message)
+
+    def log_serial(self, message: str, _: Dict[str, str] = {}) -> None:
+        if self.enable_serial_logs:
+            print(Style.DIM + message + Style.RESET_ALL)
+
+    __call__ = log_machinestate
+
+    @contextmanager
+    def nested(self, message: str) -> Iterator[None]:
+        self.log_machinestate(message)
+        yield
+
+
+class XmlLogger(Logger):
+    def __init__(self) -> None:
+        Logger.__init__(self)
+
+        self.logfile = os.environ.get("LOGFILE")
+        if self.logfile is None:
+            raise Exception(
+                "Using XmlLogger, but LOGFILE env variable not defined")
+        self.logfile_handle = codecs.open(self.logfile, "wb")
+        self.xml = XMLGenerator(self.logfile_handle, encoding="utf-8")
+        self.queue: "Queue[Tuple[str,Dict[str, str]]]" = Queue()
+
+        self.xml.startDocument()
+        self.xml.startElement("logfile", attrs={})
+
+    def release(self) -> None:
+        self.xml.endElement("logfile")
+        self.xml.endDocument()
+        self.logfile_handle.close()
+
+    def _sanitise(self, message: str) -> str:
+        return "".join(ch for ch in message if unicodedata.category(ch)[0] != "C")
+
+    def _log_line(self, message: str, attributes: Dict[str, str]) -> None:
+        self.xml.startElement("line", attributes)
+        self.xml.characters(message)
+        self.xml.endElement("line")
+
+    def log_machinestate(self, message: str, attributes: Dict[str, str] = {}) -> None:
+        common.eprint(message, attributes)
+        self._drain_log_queue()
+        self._log_line(message, attributes)
+
+    def log_serial(self, message: str, attributes: Dict[str, str] = {}) -> None:
+        self._enqueue((message, {"type": "serial"}))
+        if self.enable_serial_logs:
+            common.print(Style.DIM + message + Style.RESET_ALL)
```

Comment:
not sure if the XML logger should have to do anything with printing.

it would be very nice to be able to remove anything that has to do with printing from this class, and then have a `CompositeLogger` class that contains a normal logger and an xml logger and simply forwards calls to both. not sure how to handle the nesting for those classes but i bet there are possibilities.

Reply:
Good question!

The printing here is done to stderr instead of stdout.

I grasped it as when this is run on CI, we then want to process structured logs fr xml, but also see the logs.

I had a faint association of "CI must be stderr", but I coudn't get a proper hold on this memory any more.

Maybe you got more insight?

Reply:
Can we bounce ideas in https://matrix.to/#/#nixpkgs-pr-125372:matrix.org ? Resolving in the meantime, since I don't plan to work on this until further input / brainstorming.